File: web/typescript/gameassets/room/vector2.ts
function lerpVector2(v1: Vector2, v2: Vector2, a: number){
    return new Vector2(
        v1.x + a * (v2.x - v1.x), 
        v1.y + a * (v2.y - v1.y)
    )
}

class Vector2 {
    x: number;
    y: number;

    constructor(x: number, y: number){
        this.x = x;
        this.y = y;
    }

    setPosition(x: number, y: number){
        this.x = x;
        this.y = y;
    }
}

File: games/internal/mathobjects/vector2.go
package mathobjects

type Vector2[T float64 | int] struct {
	X T `json:"x"`
	Y T `json:"y"`
}

func (v *Vector2[T]) SetPosition(x, y T) {
	v.X = x
	v.Y = y
}

func (v *Vector2[T]) Add(v2 Vector2[T]) {
	v.X += v2.X
	v.Y += v2.Y
}

func (v Vector2[T]) Scale(scalar T) Vector2[T] {
	return Vector2[T]{
		X: v.X * scalar,
		Y: v.Y * scalar,
	}
}
